### PR TITLE
5190 endpoint to create response requisition lines

### LINKED ISSUE #5190: Service and endpoint to create response requisition lines
## Is your feature request related to a problem? Please describe 👀
Service and endpoint to create response requisition lines. Doesn't require extra columns in requisition line

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #5190

# 👩🏻‍💻 What does this PR do?
Endpoint to insert requisition lines + update `update requisition line` to include new fields.

# 🧪 Testing

- [ ] Run tests OR call API endpoint

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [x] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

File: server/service/src/requisition_line/response_requisition_line/insert/generate.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        comment,
+    }: InsertResponseRequisitionLine,
+) -> Result<RequisitionLineRow, OutError> {
+    let mut new_requisition_line =
+        generate_requisition_lines(ctx, store_id, &requisition_row, vec![item_id])?
+            .pop()
+            .ok_or(OutError::CannotFindItemStatusForRequisitionLine)?;
+
+    new_requisition_line.id = id;
+    new_requisition_line.initial_stock_on_hand_units = their_stock_on_hand.unwrap_or(0.0);
```
Comment (Chris-Petty) [inline server/service/src/requisition_line/response_requisition_line/insert/generate.rs:29]: Hmm I think this should be saved to `available_stock_on_hand`... and `initial_stock_on_hand_units`?

image

This is a response requisition generated from an internal order from a customer store. The current SOH is saved on `stock_on_hand` (which translates to `available_stock_on_hand`)

image

This is a manual response requisition with no extra columns. Seems to have saved to both `stock_on_hand` and `Cust_prev_stock_balance`. Wait wtf is the field `previous_stock_on_hand` isn't that `Cust_prev_stock_balance`?!

image

OK nvm that field is basically not used lol

File: server/service/src/requisition_line/response_requisition_line/insert/validate.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    if requisition_row.program_id.is_some() {
+        return Err(OutError::CannotAddItemToProgramRequisition);
+    }
+
+    if requisition_row.store_id != store_id {
+        return Err(OutError::NotThisStoreRequisition);
+    }
+
+    if requisition_row.status != RequisitionStatus::New {
```
Comment (Chris-Petty) [inline server/service/src/requisition_line/response_requisition_line/insert/validate.rs:31]: I suppose `Draft` should be allowed too? But doesn't really matter because they start at `New` anyway. Wondering if we even need draft lol.

File: server/service/src/requisition_line/response_requisition_line/insert/validate.rs
Comment (Chris-Petty) [inline server/service/src/requisition_line/response_requisition_line/insert/validate.rs:1]: noice

File: server/service/src/requisition_line/response_requisition_line/insert/mod.rs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        }
+    }
+
+    fn new_request_requisition() -> RequisitionRow {
+        RequisitionRow {
+            id: "draft_request_requisition".to_string(),
+            store_id: mock_store_a().id,
+            name_link_id: mock_name_b().id,
+            r#type: RequisitionType::Request,
+            status: RequisitionStatus::New,
```
Comment (Chris-Petty) [inline server/service/src/requisition_line/response_requisition_line/insert/mod.rs:101]: id says draft but status says new?

File: server/service/src/requisition_line/response_requisition_line/insert/mod.rs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+            .unwrap()
+            .unwrap();
+
+        assert_eq!(
+            line,
+            inline_edit(&line, |mut u| {
+                u.supply_quantity = 20.0;
+                u.requested_quantity = 10.0;
+                u.initial_stock_on_hand_units = 10.0;
+                u.available_stock_on_hand = test_item_stats::item_2_soh();
```
Comment (Chris-Petty) [inline server/service/src/requisition_line/response_requisition_line/insert/mod.rs:272]: Relating to my other comment about this field - this field is meant to be a snapshot of stock on hand of the customer when the requisition for stock was made. For an internal order it makes sense to look at item_stats as you have your stock data. When making a manual requisition for a facility or a store on another site (the latter being something rarely done), there is little to no chance knowing what their stock level is.

Long story short, relating to my other comment: this value should probably just be 10.0 to match the value in `initial_stock_on_hand`

**Review by Chris-Petty — CHANGES_REQUESTED** [review]
Sweet! Some changes for where to save their stock on hand

File: server/service/src/requisition_line/response_requisition_line/insert/validate.rs (hunk 2 above)
Comment (roxy-dao) [reply server/service/src/requisition_line/response_requisition_line/insert/validate.rs:31]: There shouldn't be any response requisitions with Draft 🤔 and frontend doesn't allow editing of Draft response requisitions, so frontend will need changes to if we account for it (even though they shouldn't be able to create draft response requisitions

File: server/service/src/requisition_line/response_requisition_line/insert/mod.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/requisition_line/response_requisition_line/insert/mod.rs:101]: wym?

File: server/service/src/requisition_line/response_requisition_line/insert/generate.rs (hunk 1 above)
Comment (roxy-dao) [reply server/service/src/requisition_line/response_requisition_line/insert/generate.rs:29]: I was going to say that it would overwrite the generated stock on hand for OMS customers, but then realised we get those from the linked requisition 👀 [e52d035](https://github.com/msupply-foundation/open-msupply/pull/5199/commits/e52d035357b1d8773750c1c15e39de6a214f9584)

File: server/service/src/requisition_line/response_requisition_line/insert/validate.rs (hunk 2 above)
Comment (roxy-dao) [reply server/service/src/requisition_line/response_requisition_line/insert/validate.rs:31]: Yeah no I totally agree!

**Review by Chris-Petty — APPROVED** [review]
LGTM!
